Aesthetic Changes wrt 'size_t' data types.

kanodiaayush opened the pull request:
1. Do not set or compare size_t with negative values
2. Leaf nodes are checked by testing whether their children are nullptr

Fixes #10

kanodiaayush committed:
Do not set or compare size_t with negative values; leaf nodes are checked by testing whether their children are nullptr

erikcs commented on src/tree_search.cpp:
```diff
@@ -341,15 +341,15 @@ std::unique_ptr<Node> find_best_split(const std::vector<flat_set>& sorted_sets,
         double leaf_reward = best_left_child->reward + best_right_child->reward;
         size_t leaf_action = best_left_child->action_id;
-        best_ans_as_leaf = std::unique_ptr<Node> (new Node(-1, -1, leaf_reward, leaf_action));
+        best_ans_as_leaf = std::unique_ptr<Node> (new Node(0, -1, leaf_reward, leaf_action));
```
There are still other places to make this change: like :195

erikcs replied:
And please add a more descriptive PR title and description, e.g. that this is an "aesthetic" fix

kanodiaayush replied:
Upcoming..

erikcs commented on src/tree_search.cpp:
```diff
@@ -341,15 +341,15 @@ std::unique_ptr<Node> find_best_split(const std::vector<flat_set>& sorted_sets,
     } else {
           // "pruning", the recursive case (same action in both leaves):
-      if ((best_left_child->action_id != -1) &&
+      if ((best_left_child->left_child == nullptr && best_right_child->left_child == nullptr) &&
```
And can you make a note of this in the Node struct in tree_search.h? That the convention is to set them to zero...

And what about making left_child == nullptr a method of Node? (bool is_leaf)

kanodiaayush replied:
I did consider is_leaf and I think that is a great idea. I'll do that.

kanodiaayush committed:
add is_leaf(); fix defaults for split val and split var

Nothing now defaults to -1; split val defaults to 0.0 and split var to
0; other minor consistency changes

erikcs reviewed (approved):
Thanks, looks good! I thought functions defined in the class template were inlined automatically. And the PR post is still outdated: it is in quotes and referring to 8 and 9...?

kanodiaayush commented:
> Thanks, looks good! I thought functions defined in the class template were inlined automatically. And the PR post is still outdated: it is in quotes and referring to 8 and 9...?

Fixed! I'm not sure the inline is automatic - maybe quote a reference? I'm looking here -- [here](https://stackoverflow.com/questions/10535667/does-it-make-any-sense-to-use-inline-keyword-with-templates)

erikcs commented:
https://isocpp.org/wiki/faq/inline-functions#inline-member-fns-more

kanodiaayush committed:
Remove inline from function def as defs inside class declarations are default inline

kanodiaayush commented:
Thanks, I was looking at something incorrect. Merged.
